### DISCUSSION TITLE
Gradio 4

### DISCUSSION THREAD

**Discussion by AlUlkesh** [AlUlkesh/stable-diffusion-webui-images-browser#245] [Announcements]
As you might be aware, a1111 is working on supporting the next major Gradio release.

Gradio 4 offers _custom components_, so I plan to create an extended Gallery component.

One of the main advantages is that much less custom javascript code is needed. This has been the major problem with this extension. The standard Gradio Gallery just didn't have enough features, so a lot of stuff needed to be done in javascript, which essentially tried to force Gradio to do things it wasn’t designed to do. This made it hard to develop and test, and it wasn’t always stable.

Another advantage will be that the custom component can have features that were not reasonably possible before. A good example is the ability to select multiple image and perform some kind of mass processing on them (for example delete).

Here's my current dev version and you can see the available select boxes on each thumbnail:

![image](https://github.com/AlUlkesh/stable-diffusion-webui-images-browser/assets/99896447/a3ca59f9-55b9-4c77-840c-bcef9a99a774)

This will probably be the biggest change since this extension was created, so I hope I don't put too many new bugs in :sweat_smile:

I do plan on supporting both the Gradio 3 and Gradio 4 versions for a while, if I can manage that. But at some point I'll have to switch to Gradio 4 only.

**Comment by AlUlkesh** [comment]
Short update...

I had expected the custom component to be faster than the old implementation - but this is definitely exceeding my expectations.

In the video, you can see a side-by-side comparison. As you can see, when you switch quickly from image to image, the old implementation is somewhat choppy and the generation info / filename info lags behind.

By contrast, the new implementation looks smooth and the textboxes are updated instantly.

https://github.com/AlUlkesh/stable-diffusion-webui-images-browser/assets/99896447/d15bd6e7-468a-41de-ade6-a9c27d772f08

**Reply by RamonGuthrie** [reply]
These developments are looking amazing! When I  use ComfyUI I render outputs to my Automatic1111 directory just to the **Stable-diffusion-Webui-Images-Browser** to browse images. The select boxes on each thumbnail is going to be super to review and delete more then image at a time.

**Comment by AlUlkesh** [comment]
Update: There is no update :disappointed:

a1111 still has not upgraded to Gradio 4. Gradio 4 came out last November. At this point I wonder if the switch will ever happen. This was the last info I found:

> Probably when they fix the bug with image inputs. Gradio 4 is a sad affair. It very likely will break extensions big time. Ooba updated to 4 and then reverted.

Meanwhile Gradio keeps developing, which means I feel like I'm developing against a moving target.

At this point I feel, I can't really continue, until there is more info from the a1111 side.

**Comment by RamonGuthrie** [comment]
what about supporting forge?
